README.md
## svg2Font

Converts SVG to TTF/EOT/WOFF/WOFF2/SVG format.
> Note: svg -> svgfont -> ttf  ttf -> EOT ttf -> WOFF ttf -> WOFF2 ttf -> svg，[DominantBaselineProperty](https://www.w3.org/TR/SVG/text.html#DominantBaselineProperty), [FontElementAscentAttribute](https://www.w3.org/TR/1999/WD-SVG-19991203/fonts.html#FontElementAscentAttribute)

## Unicode字符平面映射

|        平面         |     始末字符值      |                           中文名称                           |                     英文名称                     |
| :-----------------: | :-----------------: | :----------------------------------------------------------: | :----------------------------------------------: |
|       0号平面       |   U+0000 - U+FFFF   |                      **基本多文种平面**                      |      Basic Multilingual Plane，简称**BMP**       |
|       1号平面       |  U+10000 - U+1FFFF  |                      **多文种补充平面**                      |  Supplementary Multilingual Plane，简称**SMP**   |
|       2号平面       |  U+20000 - U+2FFFF  |                     **表意文字补充平面**                     |   Supplementary Ideographic Plane，简称**SIP**   |
|       3号平面       |  U+30000 - U+3FFFF  | **表意文字第三平面**（未正式使用[[1\]](https://zh.wikipedia.org/wiki/Unicode字符平面映射#cite_note-1)） |     Tertiary Ideographic Plane，简称**TIP**      |
| 4号平面 至 13号平面 |  U+40000 - U+DFFFF  |                         （尚未使用）                         |                                                  |
|      14号平面       |  U+E0000 - U+EFFFF  |                     **特别用途补充平面**                     | Supplementary Special-purpose Plane，简称**SSP** |
|      15号平面       |  U+F0000 - U+FFFFF  | 保留作为**私人使用区（A区）**[[2\]](https://zh.wikipedia.org/wiki/Unicode字符平面映射#cite_note-PUA-2) |        Private Use Area-A，简称**PUA-A**         |
|      16号平面       | U+100000 - U+10FFFF | 保留作为**私人使用区（B区）**[[2\]](https://zh.wikipedia.org/wiki/Unicode字符平面映射#cite_note-PUA-2) |        Private Use Area-B，简称**PUA-B**         |

> Unicode 字符三个私人使用区(Private Use Areas)：一个在基本多语言平面（U+E000-U+F8FF）中，另外两个几乎包含了整个第15和第16平面（分别为U+F0000-U+FFFFD，U+100000-U+10FFFD）

> ie下面渲染不出unicode，使用16进制unicode的时候 加上;号，否则ie无法将它解析。10进制的没这个问题，现在大部分字体都是使用16进制标志的，主要是为了方便移动端，还有css3的那种写法。

## Install 

```
npm install svg-to-fonts
```

## Usage

```
const { svg2Font, Font, writeToFile } = require('svg-to-fonts')
const path = require('path')
const fs = require('fs')

svg2Font({
  src: './test/svgicon/*.svg', // svg path  support patterns
  dist: './test/',
  fontFamily: 'svg2font',  // fontFamily name
  fontName: 'svg2font',  // font name
  startCodePoint: 57344, // unicode start code point
  ascent: 924,
  descent: -100,
  css: true,
  symbol: true,
  html: true,
}).then(() => {
  console.log('svg2Font done !')
})

const icon = fs.readFileSync(path.join( './test/svgicon', `icon_7days.svg`), 'utf8')

const font = new Font({
  fontFamily: 'svg2font',  // fontFamily name
  fontName: 'svgfont',
  glyphSvgs: icon,  //  support string array object
  ascent: 896,
  descent: -128,
  startCodePoint: 57344,
  css: true,
  symbol: true,
  html: true,
})

font.convertFonts({
  dist: './test/',
  fontTypes: ['eot', 'ttf', 'svg'],
  css: true,
})

```
## API

### svg2Font(options)
- `options`
  - `src`- SVG file 
  - `dist` - out file
  - `fontFamily` - font family name
  - `fontName` - font name
  - `startCodePoint` - Unicode Private Use Areas start Code Point
  - `customUnicodeList`
  - `ascent`
  - `descent`
  - `fontCdnUrl`      
  - `css`
  - `symbol`
  - `html`

### new Font(options)
- `options`
  - `fontFamily` - font family name
  - `fontName` - font name
  - `glyphSvgs` - svg datas support string array object
  - `startCodePoint` - Unicode Private Use Areas start Code Point
  - `customUnicodeList`
  - `ascent`
  - `descent`


- `method`
  - `getGlyph`
  - `getTTF`
  - `getWOFF`
  - `getWOFF2`
  - `convertFonts(options)`
    - `options`
      - `dist`
      - `fontTypes`
      - `fontCdnUrl`     
      - `css`
      - `symbol`
      - `html`  


## Features

- 支持解析 SVG 基本路径转换
- 支持导出四种浏览器主流字体（ttf，eot，woff，woff2, svg）
- 支持设置各种字体相关内容

## License

[MIT](./LICENSE)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.1.13](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.12...v0.1.13) (2020-07-23)


### Features

* add coverage ([de13c63](https://github.com/convertSvg/svg-to-fonts/commit/de13c6377bbbe2ebfc841d88abd102b46e4da1ed))
* add jest test ([42573bd](https://github.com/convertSvg/svg-to-fonts/commit/42573bd66ef2d9495d187e2f552d341a883db0c5))


### Bug Fixes

* **template:** 修复 iOS 头文件 Unicode 编码模板问题 ([229a01c](https://github.com/convertSvg/svg-to-fonts/commit/229a01cabdda9aad9a7f2c7b4e8943875f0df508))

### [0.1.12](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.11...v0.1.12) (2020-05-09)


### Bug Fixes

* **template:** fix object-c template error ([a9f5fc8](https://github.com/convertSvg/svg-to-fonts/commit/a9f5fc82aa913b28446454d122d468467c622b98))

### [0.1.11](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.10...v0.1.11) (2020-05-08)


### Bug Fixes

* **template:** fix object-c template error ([62a1bdc](https://github.com/convertSvg/svg-to-fonts/commit/62a1bdc3a694a9fd75a144dc0c16a9d52289eaa0))

### [0.1.10](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.9...v0.1.10) (2020-04-07)


### Features

* **xml:** fixed xml comment ([bf82a6c](https://github.com/convertSvg/svg-to-fonts/commit/bf82a6c4635814bf4398a0daf12a90d203e01781))

### [0.1.9](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.8...v0.1.9) (2020-04-07)


### Features

* **rn:** 增加 rn 模版列表文件 ([a39fa56](https://github.com/convertSvg/svg-to-fonts/commit/a39fa56269b43eba635bb0d8ea760b2dd0695fee))

### [0.1.8](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.6...v0.1.8) (2020-04-07)

### [0.1.6](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.5...v0.1.6) (2020-04-07)


### Features

* **font:** add AndroidTemplate iOSTemplate ([e2436be](https://github.com/convertSvg/svg-to-fonts/commit/e2436be99e20c336f019ffda3dab374a4f190891))

### [0.1.5](https://github.com/convertSvg/svg-to-fonts/compare/v0.1.4...v0.1.5) (2020-03-19)

### 0.1.4 (2020-03-19)


### Features

*  完善 svg2Font ([22c0d33](https://github.com/convertSvg/svg-to-fonts/commit/22c0d33df65ec15f0edb79731b2668cf0ec05bea))
* add dist file ([a7354a3](https://github.com/convertSvg/svg-to-fonts/commit/a7354a3ddf065722a55965503e241524b1e7193d))
* add font css feature ([61406f0](https://github.com/convertSvg/svg-to-fonts/commit/61406f07edf1c1e93a569a76cc5b97522e865e35))
* add fontCdnUrl ([39f5739](https://github.com/convertSvg/svg-to-fonts/commit/39f57390d00b1b70a605de1acb459c04ecebc543))
* main process ([c5fab43](https://github.com/convertSvg/svg-to-fonts/commit/c5fab431951b2452ff19480739d662afdf465658))
* modify package.json and publish ([2932f21](https://github.com/convertSvg/svg-to-fonts/commit/2932f2130c9af5587063805f57ee4177f1fe645a))
* support custom unicodes and fontclass ([e17a33f](https://github.com/convertSvg/svg-to-fonts/commit/e17a33f19705c708a45954d0f53c3dddf21fc9a6))
* support fontcss symbol ([a159812](https://github.com/convertSvg/svg-to-fonts/commit/a15981240219aec43edd6e858a996cb60bd998b1))
* support multiple paths ([f424313](https://github.com/convertSvg/svg-to-fonts/commit/f42431382faa825e2f64fd38b3a3c757b82df47a))
* update package.json readme ([71b59fe](https://github.com/convertSvg/svg-to-fonts/commit/71b59fef2830369de75fa45451e4cca1e374cb7d))
* update readme and publish npm package ([7604870](https://github.com/convertSvg/svg-to-fonts/commit/7604870fdfd937221265efe2ba2c59eedaa8c931))
* 支持自定义 fontfamilyclass ([c864ddd](https://github.com/convertSvg/svg-to-fonts/commit/c864ddddb141dc27d929d87e173b0af262405ac3))
* 调整 rollup 打包 ([62dce40](https://github.com/convertSvg/svg-to-fonts/commit/62dce409b571248eede5985e74eccf1f20238831))


### Bug Fixes

* modify template ([1846333](https://github.com/convertSvg/svg-to-fonts/commit/1846333fed4bcd702d80efba16927dd5c732dd9b))
* update gitignore ([634abbb](https://github.com/convertSvg/svg-to-fonts/commit/634abbb2388c12ecdf133684d65e69c084bd63eb))
* WOFF error ([08cd6ff](https://github.com/convertSvg/svg-to-fonts/commit/08cd6ff96d12b8858c35205f119379dcbab27bbd))
* 修正模板 eot 字体引入路径 ([8370c48](https://github.com/convertSvg/svg-to-fonts/commit/8370c4837fb3ab391d749e7a9cecd2f47830ad89))


test/svg2font.js

  (function (document) {
var symbols = '<svg xmlns="http://www.w3.org/2000/svg" style="display:none"><symbol id="icon_7days" viewBox="0 0 1024 1024"><path d="M511.96 1023.97c-282.76 0-511.96-229.23-511.96-511.98 0-282.71 229.21-511.94 511.96-511.94 282.73 0 511.96 229.23 511.96 511.94 0 282.76-229.23 511.99-511.96 511.98z m0-973.03c-254.62 0-461.05 206.42-461.05 461.05s206.42 461.05 461.05 461.04 461.02-206.42 461.02-461.04-206.4-461.05-461.02-461.05z m-97.27 768.28c23.83-171.38 68.5-338.28 199.66-537.59h-307.17v-51.2h383.97v32.72c-137 199.33-207.52 376.55-230.08 555.91-15.67 0.15-46.38 0.15-46.38 0.16z" fill="#6A79B4"></path></symbol><symbol id="icon_hour" viewBox="0 0 1024 1024"><path d="M971.55 361.72h-159.84v-50.35h109.47a39.96 39.96 0 0 0 39.96-39.96v-168.61a39.96 39.96 0 0 0-39.96-39.94h-818.36a39.96 39.96 0 0 0-39.96 39.94v168.61a39.96 39.96 0 0 0 39.96 39.96h84.5v50.35h-134.87a39.96 39.96 0 0 1-39.96-39.96v-269.31a39.96 39.96 0 0 1 39.96-39.96h919.1a39.96 39.96 0 0 1 39.96 39.96v269.31a39.96 39.96 0 0 1-39.96 39.96z m-284.67 335.7c0 21.53-17.81 39.01-39.79 39.01h-269.78c-21.95 0-39.79-17.48-39.79-39.01v-85.84h49.63v39.01c0 21.53 17.81 36.66 39.78 36.66h169.51c21.98 0 39.79-17.46 39.79-39.01v-110.89h-248.96c-27.47 0-49.75-22.3-49.75-49.83v-175.15c0-27.52 22.28-49.8 49.75-49.8h249.88c27.47 0 49.73 22.28 49.73 49.8v175.15c0 8.82-1.6 16.98-4.45 24.18h4.45v185.72z m-50-360.25a24.98 24.98 0 0 0-24.98-24.97h-199.8a24.98 24.98 0 0 0-24.98 24.97v124.88a24.98 24.98 0 0 0 24.98 24.97h199.8a24.98 24.98 0 0 0 24.98-24.97v-124.88z m-499.74 584.83a39.96 39.96 0 0 0 40.09 39.91h669.29c22.15 0 40.11-17.86 40.11-39.91v-559.85h49.66v609.4c0 22.08-17.83 39.96-39.84 39.96h-768.9a39.89 39.89 0 0 1-39.84-39.96v-609.4h49.43v559.85z" fill=""></path></symbol></svg>'
    document.body.insertAdjacentHTML('afterBegin', symbols)
  })(document);

test/JDIF_Iconfont.h

  //
  // JDIF_Iconfont.h
  // fontFamily:iconfont
  //
  // Version 1.0
  // Copyright (C) 2019 by original authors @ master Gao
  //

  #ifndef JDIF_Iconfont_H
  #define JDIF_Iconfont_H

  NSString * const JDIF_Iconfont = @"iconfont";

  NSString * const JDIF_ICON_7DAYS = @"\U0000e975"; // icon_7days1
  NSString * const JDIF_ICON_HOUR = @"\U0000e973"; // icon_7days1

  #endif

test/icon_iconfont.js
export default {
  'icon_7days': 59765,// icon_7days1
  'icon_hour': 59763,// icon_7days1
}

index.js
const { svg2Font, Font, writeToFile } = require('./dist/index.js')
const path = require('path')
const fs = require('fs')

// svg2Font({
//   src: './test/svgicon/*.svg', // svg path  support patterns
//   dist: './test/dist/',
//   fontName: 'svg2font',  // font name
//   startCodePoint: 57344, // unicode start code point
//   ascent: 924,
//   descent: -100,
//   css: true,
//   symbol: true,
// }).then(() => {
//   console.log('svg2Font done !')
// })

const icon = fs.readFileSync(path.join( './test/svgicon', `icon_7days.svg`), 'utf8')
const icon1 = fs.readFileSync(path.join( './test/svgicon', `icon_postage.svg`), 'utf8')
console.error('icon', icon)
const font = new Font({
  fontName: 'svg2font',
  fontFamily: 'iconfont',
  fontFamilyClass: 'iconfont',
  // glyphSvgs: [icon, icon1],  // string array object
  glyphSvgs: {'icon_7days': { path: icon, originName: 'icon_7days1' }, 'icon_hour': { path: icon1 , originName: 'icon_7days1' }},  // string array object
  ascent: 896,
  descent: -128,
  startCodePoint: 57344,
  customUnicodeList: ['e975', 'e973'] // custom defined codePoint 57344 ~ 63743
})

font.convertFonts({
  dist: './test/',
  fontTypes: ['eot', 'woff2', 'woff', 'ttf', 'svg'],
  fontCdnUrl: '//storage.360buyimg.com/lcstatic/plugins/',
  css: true,
  symbol: true,
  html: true,
})


// const svgFont = font.getGlyph()

// writeToFile(svgFont, path.join( './test/', `svg2font.svg`))

// const svgFontTTF = font.getTTF()

// fs.writeFileSync(path.join( './test/', `svg2font.ttf`), svgFontTTF)


// const svgFontEOT = font.getEOT()

// fs.writeFileSync(path.join( './test/', `svg2font.eot`), svgFontEOT)

// const svgFontWOFF = font.getWOFF()
// fs.writeFileSync(path.join( './test/', `svg2font.woff`), svgFontWOFF)


// const svgFontWOFF2 = font.getWOFF2()
// fs.writeFileSync(path.join( './test/', `svg2font.woff2`), svgFontWOFF2)
